keps/sig-windows/0000-20181221-windows-group-managed-service-accounts-for-container-identity.md

```diff
@@ -88,119 +93,243 @@ To achieve the scale and speed required for containers, Windows uses a group man
+  - Create a cluster wide role to get and "use" the GMSACredSpec:
 
 ```
-kind: Role
 apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
 metadata:
-  namespace: default
-  name: webserver-gmsa-user
+  name: webapp1-gmsa-user
 rules:
-- apiGroups: [""]
-  resources: ["configmaps"]
-  resourceNames: ["webserver-credspec"]
-  verbs: ["get", "patch", "delete", "update", "use"]
+- apiGroups: ["auth.k8s.io"]
+  resources: ["gmsacredspecs"]
+  resourceNames: ["webapp1-credspec"]
+  verbs: ["get", "use"]
 ```
 
-  - Create a rolebinding and assign desired users (for example, jane below) or service accounts to the above role:
+  - Create a rolebinding and assign desired service accounts to the above role:
 
 ```
-kind: RoleBinding
 apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
 metadata:
-  name: use-webserver-gmsa
+  name: use-webapp1-gmsa
   namespace: default
 subjects:
-- kind: User
-  name: jane
+- kind: ServiceAccount
+  name: webapp-sa
   apiGroup: rbac.authorization.k8s.io
 roleRef:
   kind: Role
-  name: webserver-gmsa-user
+  name: webapp1-gmsa-user
   apiGroup: rbac.authorization.k8s.io
 ```
   
-4. Application admins deploy app pods with a reference to the name of an authorized GMSA configmap (e.g. `webserver-credspec`). In the alpha and beta stage of this feature, the reference to the GMSA configmap will be set through an annotation on the pod: `pod.[alpha|beta].kubernetes.io/windows-gmsa-config-map`. In the beta and GA stage, the reference to GMSA configmap may be set through `spec.securityContext.windows.credentialSpecConfig` in the pod spec.
-5. During pod creation, a new Admission Controller, GMSAAuthorizer, will ensure the `user` (and `spec.serviceAccountName` if specified) is authorized for the `use` verb on the GMSA configmap. If the authorization check fails due to absence of RBAC roles, the pod creation fails.
-6. The Windows CRI implementation accesses the configmap with the GMSA credspec and copies the contents of entry with key `gmsa-cred-spec` to the [OCI windows.CredentialSpec](https://github.com/opencontainers/runtime-spec/blob/master/config-windows.md#credential-spec) field.
-7. The Windows OCI implementation validates the credspec and fails to start the container if its invalid or access to the GMSA from the node is denied.
-8. The container starts with the Windows Group Managed Service Account as expected, and can authenticate to the database.
-9. During any subsequent pod update, any changes to a pod's `securityContext` is blocked (as is the case today) by `ValidatePodUpdate`.
+5. Application admins will deploy app pods that require a GMSA identity along with a Service Account authorized to use the GMSAs. There will be two ways to specify the GMSA credspec details for pods and containers. It is expected that users will typically use the first option as it is more user friendly. The second option is available mainly due to an artifact of the design choices made and described here for completeness.
```

Comment:
> It is expected that users will typically use the first option as it is more user friendly

if the admission plugin allows direct specification, doesn't that allow bypass of the authz check? I'd expect the admission plugin to override any user-specified credspec annotations on pod creation

Reply:
@liggitt we specifically have a check for credspec specified in expanded form by the user. It is covered in step [6] as part of the actions the webhook performs:

`If a credspec JSON is specified, look up the corresponding GMSACredSpec name and perform a deep equal check of the members of the specified credspec and prepared credspec from previous step. If the deep equal check fails, GMSAAuthorizer will be fail pod creation with 400: BadRequest.`

Note also that we specify that a name referring to a GMSACredSpec is mandatory (regardless of whether the expanded contents is specified so that the name can be validated against): 

`If an application administrator wants containers to be initialized with a GMSA identity, specifying the names of the desired GMSACredSpec objects is mandatory
`

Reply:
ah, hadn't gotten down to that yet. thanks.